From 0228341cee2bd960b58db9a5e5569de59950dc16 Mon Sep 17 00:00:00 2001
From: Janne Grunau <j@jannau.net>
Date: Tue, 21 Jan 2025 21:04:33 +0100
Subject: [PATCH 49/64] HID: transport: spi: apple: Use distinctive names input
 devices

Kwin uses the tripple Vendor ID, Product ID, Name to identify input
devices. If the same name ("Apple Internal Keyboard / Trackpad") is used
for both keyboard and trackpad actions like disabling the touchpad have
confusing effects. On the next login the keyboard will be disabled as
well since it shares all identifiers.

Signed-off-by: Janne Grunau <j@jannau.net>
---
 drivers/hid/spi-hid/spi-hid-apple-core.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/hid/spi-hid/spi-hid-apple-core.c b/drivers/hid/spi-hid/spi-hid-apple-core.c
index aff822d61..1f8fa64d6 100644
--- a/drivers/hid/spi-hid/spi-hid-apple-core.c
+++ b/drivers/hid/spi-hid/spi-hid-apple-core.c
@@ -903,6 +903,7 @@ static int spihid_register_hid_device(struct spihid_apple *spihid,
 				      struct spihid_interface *iface, u8 device)
 {
 	int ret;
+	char *suffix;
 	struct hid_device *hid;
 
 	iface->id = device;
@@ -911,7 +912,16 @@ static int spihid_register_hid_device(struct spihid_apple *spihid,
 	if (IS_ERR(hid))
 		return PTR_ERR(hid);
 
-	strscpy(hid->name, spihid->product, sizeof(hid->name));
+	/*
+	 * Use 'Apple SPI Keyboard' and 'Apple SPI Trackpad' as input device
+	 * names. The device names need to be distinct since at least Kwin uses
+	 * the tripple Vendor ID, Product ID, Name to identify devices.
+	 */
+	snprintf(hid->name, sizeof(hid->name), "Apple SPI %s", iface->name);
+	// strip ' / Boot' suffix from the name
+	suffix = strstr(hid->name, " / Boot");
+	if (suffix)
+		suffix[0] = '\0';
 	snprintf(hid->phys, sizeof(hid->phys), "%s (%hhx)",
 		 dev_name(&spihid->spidev->dev), device);
 	strscpy(hid->uniq, spihid->serial, sizeof(hid->uniq));
-- 
2.50.0

